### ISSUE TITLE
2.4.0 breaks template extraction of git urls

### ISSUE THREAD

**Issue by errm** [sporkmonger/addressable#234]
I was doing something like this in one of my apps to parse github ssh urls...

``` ruby
require "addressable/template"

template = Addressable::Template.new('{user}@{host}:{org}/{repo}.git')
uri = Addressable::URI.parse("git@github.com:awesome/sauce.git")
puts  template.extract(uri).inspect
```

On `2.3.8` this works as expected and I get my expected output

``` ruby
{"user"=>"git", "host"=>"github.com", "org"=>"awesome", "repo"=>"sauce"}
```

When I upgraded to `2.4.0` the call to `Addressable::URI.parse` blows up in all sorts of fire:

```
/usr/local/var/rbenv/versions/2.1.8/lib/ruby/gems/2.1.0/gems/addressable-2.4.0/lib/addressable/uri.rb:867:in `scheme=': Invalid scheme format: git@github.com (Addressable::URI::InvalidURIError)
    from /usr/local/var/rbenv/versions/2.1.8/lib/ruby/gems/2.1.0/gems/addressable-2.4.0/lib/addressable/uri.rb:795:in `block in initialize'
    from /usr/local/var/rbenv/versions/2.1.8/lib/ruby/gems/2.1.0/gems/addressable-2.4.0/lib/addressable/uri.rb:2302:in `call'
    from /usr/local/var/rbenv/versions/2.1.8/lib/ruby/gems/2.1.0/gems/addressable-2.4.0/lib/addressable/uri.rb:2302:in `defer_validation'
    from /usr/local/var/rbenv/versions/2.1.8/lib/ruby/gems/2.1.0/gems/addressable-2.4.0/lib/addressable/uri.rb:792:in `initialize'
    from /usr/local/var/rbenv/versions/2.1.8/lib/ruby/gems/2.1.0/gems/addressable-2.4.0/lib/addressable/uri.rb:135:in `new'
    from /usr/local/var/rbenv/versions/2.1.8/lib/ruby/gems/2.1.0/gems/addressable-2.4.0/lib/addressable/uri.rb:135:in `parse'
```

Was this an intentional change? I understand that the thing I am parsing is not exactly a valid URI, but in the past addressable was quite liberal about these things.

Would there be (or is there) a way to skip validating the url when parsing stuff out of possibly invalid urls using templates?

**Comment by sporkmonger** [comment]
It's not a URI, behavior when calling `parse` is therefore undefined. That said, I've considered adding support for parsing URI-like things in the `three` branch, which has a design that lends itself more towards this. I'm still on the fence about it, but Addressable does parse URI templates, which are not themselves URIs, so SCP references could certainly theoretically fall under Addressable's purvue. They just wouldn't be handled by `Addressable::URI`.

**Comment by connorshea** [comment]
We just ran into that at GitLab as well: https://gitlab.com/gitlab-org/gitlab-ce/merge_requests/5250

**Comment by sbamin** [comment]
I find failures in building github hosted sites as current github pages-gem v127 requires addressable v2.4 (via jekyll 3.4.1 dependency), and throwing the same error as above. See https://github.com/github/pages-gem/issues/424

`bundle exec jekyll build --safe` works without any error locally if I downgrade addressable to v2.3.8 (and also pages-gem to v99). Of course, github pages uses most current version of pages-gem with addressable v2.4, and hence, site fails to build on the github server.

**Comment by sporkmonger** [comment]
Good to know, but it's their bug, not Addressable's. I'm closing this as "Won't Fix", but might revisit for 3.0.

**Comment by sbamin** [comment]
weird as github manager, @benbalter thinks the same way! I will leave it closed on both end, and instead will wait until the issue get picked by others.

https://github.com/github/pages-gem/issues/424
